Utils/Nginx/config.h

```diff
@@ -0,0 +1,207 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_UTILS_NGINX_CONFIG_H
+#define CURRENT_UTILS_NGINX_CONFIG_H
+
+#include "../../port.h"
+
+#include "exceptions.h"
+
+#include "../../Bricks/strings/util.h"
+#include "../../TypeSystem/struct.h"
+
+namespace current {
+namespace utils {
+namespace nginx {
+namespace config {
+
+CURRENT_STRUCT(SimpleDirective) {
+  CURRENT_FIELD(name, std::string);
+  CURRENT_FIELD(parameters, std::string);
+  CURRENT_DEFAULT_CONSTRUCTOR(SimpleDirective){};
+  CURRENT_CONSTRUCTOR(SimpleDirective)(const std::string& name, const std::string& parameters)
+      : name(name), parameters(parameters) {}
+};
+
+CURRENT_FORWARD_DECLARE_STRUCT(BlockDirective);
+CURRENT_FORWARD_DECLARE_STRUCT(ServerDirective);
+CURRENT_FORWARD_DECLARE_STRUCT(LocationDirective);
+
+using directives_t = Variant<SimpleDirective, BlockDirective, ServerDirective, LocationDirective>;
+
+CURRENT_STRUCT(BlockDirective) {
+  CURRENT_FIELD(name, std::string);
+  CURRENT_FIELD(parameters, std::string);
+  CURRENT_FIELD(directives, std::vector<directives_t>);
+  CURRENT_DEFAULT_CONSTRUCTOR(BlockDirective) {}
+  CURRENT_CONSTRUCTOR(BlockDirective)(const std::string& name) : name(name) {}
+  CURRENT_CONSTRUCTOR(BlockDirective)(const std::string& name, const std::string& parameters)
+      : name(name), parameters(parameters) {}
+
+  SimpleDirective& AddSimpleDirective(const std::string& name, const std::string& parameters) {
+    directives.push_back(SimpleDirective(name, parameters));
+    return Value<SimpleDirective>(directives.back());
+  }
+
+  template <typename T>
+  T& AddDirective(T && directive) {
+    directives.push_back(std::forward<T>(directive));
+    return Value<T>(directives.back());
+  }
+};
+
+// clang-format off
+CURRENT_STRUCT(LocationDirective, BlockDirective) {
+  CURRENT_DEFAULT_CONSTRUCTOR(LocationDirective) : SUPER("location") {}
+  CURRENT_CONSTRUCTOR(LocationDirective)(const std::string& location) : SUPER("location", location) {}
+
+  LocationDirective& AddNestedLocation(const std::string& location) {
+    return SUPER::AddDirective(LocationDirective(location));
+  }
+};
+// clang-format on
+
+inline LocationDirective DefaultLocationDirective(const std::string& location) {
+  LocationDirective result(location);
```

Comment:
Could be confusing if `AddSimpleDirective` returns a reference to a parent, since `AddBlockDirective` should return a reference to a newly created object.